### PR TITLE
JAMES-3726 Allow relaxing CassandraMailRepository LWT usage

### PR DESCRIPTION
**Pull request by vttranlina**
Jira https://issues.apache.org/jira/browse/JAMES-3726

### CODE REVIEW

File: server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java [hunk 1]
```diff
@@ -28,56 +28,91 @@
...
     }
 
     public Mono<Boolean> store(MailRepositoryUrl url, MailKey key) {
+        if (strongConsistency) {
+            return storeExecute(url, key);
+        }
+        return exists(url, key)
+            .filter(FunctionalUtils.identityPredicate())
+            .map(FunctionalUtils.negate())
+            .switchIfEmpty(storeExecute(url, key));
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java:112]: Please let's not do read - before writes. I agree having a different behaviour than when we use the strong consistency mode.

That is: always store!

File: server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java [hunk 2]
```diff
@@ -90,8 +125,24 @@ private PreparedStatement prepareInsert(Session session) {
     }
 
     public Mono<Boolean> remove(MailRepositoryUrl url, MailKey key) {
+        if (strongConsistency) {
+            return removeExecute(url, key);
+        }
+        return exists(url, key)
+            .filter(Predicate.not(FunctionalUtils.identityPredicate()))
+            .switchIfEmpty(removeExecute(url, key));
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java:133]: Idem always delete!

**Review by chibenwa — COMMENTED** [review]
DOC in `server/apps/distributed-app/docs` ?

File: server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java (hunk 1 above)
Comment (vttranlina) [reply server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java:112]: But we have a test case: `storeShouldReturnFalseWhenPreviouslyStored`
And the boolean result will change the logic of `CassandraMailRepositoryCountDAO`. 
?

Comment (chibenwa) [reply server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepositoryKeysDAO.java:112]: > storeShouldReturnFalseWhenPreviouslyStored

When I turn off LWTs I want `storeShouldReturnTrueWhenPreviouslyStored` (idempotent behaviour)

File: server/apps/cassandra-app/sample-configuration/cassandra.properties [hunk 3]
```diff
@@ -52,3 +52,7 @@ cassandra.retryConnection.minDelay=5000
 # Falls back to default read consistency level if the blob is missing.
 # Defaults to false.
 # optimistic.consistency.level.enabled=false
+
+# Optional. Allows to not using lightweight transactions in CassandraMailRepository.
+# Defaults to false.
+# mailrepository.strong.consistency=false
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/apps/cassandra-app/sample-configuration/cassandra.properties:58]: The default should be `true` ie no breaking changes

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/cassandra.adoc [hunk 4]
```diff
@@ -215,4 +215,7 @@ If disabled, Lightweight transactions will no longer be used upon messages opera
 As message flags updates relies so far on a read-before-write model, it exposes yourself to data races leading to
 potentially update loss. Better performance are expected
 by turning it off. Reads performed as part of write transaction are also performed with a relaxed consistency.
+
+| mailrepository.strong.consistency
+| Optional. Boolean, defaults to true. Allows to not using lightweight transactions in CassandraMailRepository.
```
Comment (chibenwa) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/cassandra.adoc:220]: Allows not to use lightweight transactions

Comment (chibenwa) [reply server/apps/distributed-app/docs/modules/ROOT/pages/configure/cassandra.adoc:220]: Maybe say also that we instead implement an indempotent behaviour (duplicates are overriden, missing entries upon deletes are ignored)

File: server/apps/cassandra-app/sample-configuration/cassandra.properties [hunk 5]
```diff
@@ -53,6 +53,7 @@ cassandra.retryConnection.minDelay=5000
 # Defaults to false.
 # optimistic.consistency.level.enabled=false
 
-# Optional. Allows to not using lightweight transactions in CassandraMailRepository.
+# Optional. Allows not to use lightweight transactions in CassandraMailRepository.
+# Then we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```
Comment (chibenwa) [inline server/apps/cassandra-app/sample-configuration/cassandra.properties:57]:
```suggestion
# If disabled we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/cassandra.adoc [hunk 6]
```diff
@@ -217,5 +217,6 @@ potentially update loss. Better performance are expected
 by turning it off. Reads performed as part of write transaction are also performed with a relaxed consistency.
 
 | mailrepository.strong.consistency
-| Optional. Boolean, defaults to true. Allows to not using lightweight transactions in CassandraMailRepository.
+| Optional. Boolean, defaults to true. Allows not to use lightweight transactions in CassandraMailRepository.
+| Then we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```
Comment (chibenwa) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/cassandra.adoc:221]:
```suggestion
| If disabled we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```

File: server/apps/distributed-app/sample-configuration/cassandra.properties [hunk 7]
```diff
@@ -53,6 +53,7 @@ cassandra.retryConnection.minDelay=5000
 # Defaults to false.
 # optimistic.consistency.level.enabled=false
 
-# Optional. Allows to not using lightweight transactions in CassandraMailRepository.
+# Optional. Allows not to use lightweight transactions in CassandraMailRepository.
+# Then we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```
Comment (chibenwa) [inline server/apps/distributed-app/sample-configuration/cassandra.properties:57]:
```suggestion
# If disabled we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```

File: server/apps/distributed-pop3-app/sample-configuration/cassandra.properties [hunk 8]
```diff
@@ -53,6 +53,7 @@ cassandra.retryConnection.minDelay=5000
 # Defaults to false.
 # optimistic.consistency.level.enabled=false
 
-# Optional. Allows to not using lightweight transactions in CassandraMailRepository.
+# Optional. Allows not to use lightweight transactions in CassandraMailRepository.
+# Then we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```
Comment (chibenwa) [inline server/apps/distributed-pop3-app/sample-configuration/cassandra.properties:57]:
```suggestion
# If disabled we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).
```

File: src/site/xdoc/server/config-cassandra.xml [hunk 9]
```diff
@@ -184,7 +184,8 @@
         <dd>Optional. Defaults to false. Allows specifying consistency level ONE for reads in Cassandra BlobStore.
             Falls back to default read consistency level if the blob is missing.</dd>
         <dt><strong>mailrepository.strong.consistency</strong></dt>
-        <dd>Optional. Defaults to true. Allows to not using lightweight transactions in CassandraMailRepository.</dd>
+        <dd>Optional. Defaults to true. Allows not to use lightweight transactions in CassandraMailRepository.
+            Then we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).</dd>
```
Comment (chibenwa) [inline src/site/xdoc/server/config-cassandra.xml:188]:
```suggestion
            If disabled we implement an idempotent behaviour (duplicates are overridden, missing entries upon deletes are ignored).</dd>
```

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
BTW @ottoka this might interest you a lot to avoid bad surprises upon your error handling.

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by vttranlina** [comment]
rebase master & squash fixup

**Pull request merged**